### DISCUSSION TITLE
Is Ghidra interpreting 80x86 `MOV` instruction correctly?

### DISCUSSION THREAD

**Discussion by Wall-AF** [NationalSecurityAgency/ghidra#5221] [Q&A]
Looking at the 80386 [`MOV`](https://pdos.csail.mit.edu/6.828/2005/readings/i386/MOV.htm) instruction it says the format for OpCode C7 is:
```
C7       MOV r/m16,imm16   2/2           Move immediate word to r/m word
C7       MOV r/m32,imm32   2/2           Move immediate dword to r/m dword
```
However, my disassembly is showing:
```
   1078:17b8 c7 82 62     0b4           MOV        word ptr [BP + SI + 0xff62]=>local_a0,0x0
             ff 00 00
```
which is implying an index (`SI`) into an array at the top(ish) of the stack.

Can someone confirm if my understanding is correct or have I got this completely wrong?

I'm asking because the `ModR/M` byte `82` maybe shouldn't be being interpreted as such (because there's no `/r` in th OpCode column as there is for other versions of `MOV`) and if it should then the definintion of `82` is `[BX+SI]+disp16` indicating an incorrect register use!

**Comment by Wall-AF** [comment]
Can anyone help me modify the `ia.sinc` (just locally) to change the interpretation above as I don't really understand the way it's works properly?  Also how can I adjust the assembly output to include the segment that any given instruction is using?

**Comment by GhidorahRex** [comment]
This looks correct. The only thing I'd add is that the offset is a little wonky. I think it should be - 0x9e not + 0xff62, but I'll try to look at that part carefully.

**Reply by Wall-AF** [reply]
Mmm!!!  Why do you say it's corrrect?  In the [link](https://pdos.csail.mit.edu/6.828/2005/readings/i386/MOV.htm) (and rest of site) it suggests that there's no use of an index register.

**Reply by GhidorahRex** [reply]
The addressing mode is handled as part of the modrm byte, which as you said, is 82. The description in the documentation is noted as `[BX + SI] + disp16` but it's almost always written in assemblers and disassemblers as `[BX + SI + disp16]` (see, for example https://wiki.osdev.org/X86-64_Instruction_Encoding). But a note below the table in section 2.1.5 of the Intel 64/32 Software Developer Manual says:
> 1. The disp16 nomenclature denotes a 16-bit displacement that follows the ModR/M byte and that is added to the index.

**Reply by Wall-AF** [reply]
My misunderstanding, sorry.  My large(ish) recursive function is not an easy one as there's only 50h bytes reserved for stack usage and (I'm assuming) an internal Borland function call within it (called from several places) that is performing a Ghidra indecipherable opertion to manipulate the stack which also doesn't help!

**Reply by Wall-AF** [reply]
> The addressing mode is handled as part of the modrm byte, which as you said, is 82. The description in the documentation is noted as `[BX + SI] + disp16` ...

My point is that when interpreting ModR/M byte`82` this instruction is using the **`BP`** register instead of **`BX`** which would give a completely different output.  Isn't that wrong??

**Reply by GhidorahRex** [reply]
Sorry, I was looking at the wrong column. Since the rm portion is `010` it should be `[BP + SI + disp16]`

**Reply by Wall-AF** [reply]
I'm seeing that `82` = `10 000 010` (`MOD = 10, REG/OPCODE = 000, R/M = 10`) which Table 17-2. 16-Bit Addressing Forms with the ModR/M Byte" [of this bettter link](http://intel80386.com/386htm/s17_02.htm) shows:
```
   Effective 
+---Address--+ +Mod R/M+ +--------ModR/M Values in Hexadecimal--------+
[BX+SI]+disp16   10  010   82    8A    92    9A    A2    AA    B2    BA
```

I cannot find any line representing `[BP+SI]+disp16`.

**Reply by GhidorahRex** [reply]
The updated official documentation from intel has that line as `[BP+SI]+disp16`. On the page you linked, both rm=000 and rm=010 (as well as rm=001 and rm=011) have the same formatting, so I think that's a typo from the 1986 manual. Every other page I can find a reference to the modrm bits have it as `[BP+SI]+disp16`. Since x86 protected mode is backwards compatible, I don't think it's possible that this line is incorrect in the updated documentation.

**Reply by Wall-AF** [reply]
Duh!  I see and realise now!!!  Oh well, I was hoping I'd found an issue as Ghidra is struggling with my project!  Thanks for your patience @GhidorahRex.

Btw, is there a way (presumably in the sleigh description - `ia.sinc`) to enhance the assembly to show which segment registers are being used in each mnemonic representation?  As the next stage is to get Ghidra to use them properly?

**Reply by GhidorahRex** [reply]
It's going to depend on the instruction mnemonic. Any mnemonic that accesses memory using mem16, or any instruction that calls the `segment` pcodeop is going to use a segment register. In some cases you have both with `segment(seg16,addr16)`. You would either want to add `seg16^` to the front of memory address display, or if the segment operator uses a specific segment register such as `SS` then you would want to add `SS^` to the front of whatever address field is using that.

**Reply by Wall-AF** [reply]
That's helpful, @GhidorahRex.  Thank you.

**Comment by Wall-AF** [comment]
Everything seems correct, my misunderstanding, plus other complexities made me question this!
